jax/interpreters/pxla.py

```diff
@@ -698,6 +682,23 @@ def dynamic_fun(dummy, *args):
   num_local_shards = num_local_replicas * num_partitions
   num_global_shards = num_global_replicas * num_partitions
 
+  if (xb.host_count() > 1 and
+      must_run_on_all_devices and
+      num_local_shards != xb.local_device_count()):
```

Comment:
Similarly, this check is also problematic for nested multihost pmaps.
On the same cmap/pmap example as above, we get for the inner pmap:
```
num_local_replicas = 4 * 1 = 4
num_local_shards = 4 * 1 = 4
num_local_shards != xb.local_device_count()
```

I'm not sure how to fix this chunk, but I think the reasonable thing to do is to say must_run_on_all_devices if global_axis_size is specified (and optionally is less than local_device_count).

Reply:
I expanded the conditional to only fire for outer pmaps. Everything should add up in the outer pmap. There might be errors we could check for in the inner pmap too, but this code won't be run for inner pmaps soon (see above comment).